File: src/Logging/logging.c

#include "../helper/struct.h"
#include <stdio.h>
#include <stdlib.h>

FILE* fileptr = NULL;
FILE* fileptrcsv = NULL;
FILE* fileptrconfig = NULL;

void open_file(gene_pool_t gene_pool)
{
	if(fopen_s (&fileptr, "c:/temp/param.bin", "wb") != 0 ||
	   fopen_s (&fileptrcsv, "c:/temp/param.csv", "w") != 0 ||
	   fopen_s (&fileptrconfig, "c:/temp/config.json", "w") != 0)
	{
		printf("Error opening file!\n");
		exit(1);
	}
	fprintf(fileptrcsv, "iteration, individual, result, ");
	for (int i = 0; i < gene_pool.genes; i++)
	{
		fprintf(fileptrcsv, "gene%d, ", i);
	}
	fprintf(fileptrcsv, "\n");


}

void close_file()
{
	fclose(fileptr);
	fclose(fileptrcsv);
	fclose(fileptrconfig);
}

void write_param(gene_pool_t gene_pool, int iteration)
{
	if (fileptr == NULL)
	{
		printf("Error opening file!\n");
		exit(1);
	}
	fwrite(&iteration, sizeof(int), 1, fileptr);
	fwrite(&gene_pool.pop_result_set, sizeof(gene_pool.pop_result_set), 1, fileptr);

	int paramsetsize = sizeof(gene_pool.pop_param_double[0]);

	for (int i = 0; i < gene_pool.individuals; i++)
	{
		fwrite(&gene_pool.pop_param_double[i], paramsetsize, 1, fileptr);

		fprintf(fileptrcsv, "%d, %d, %f, ", iteration, i, gene_pool.pop_result_set[gene_pool.sorted_indexes[i]]);
		for (int j = 0; j < gene_pool.genes; j++)
		{
			fprintf(fileptrcsv, "%f, ", gene_pool.pop_param_double[i][j]);
		}
		fprintf(fileptrcsv, "\n");
	}
}

void write_config(gene_pool_t gene_pool, runtime_param_t run_param, config_ga_t config_ga)
{

	if (fileptrconfig == NULL)
	{
		printf("Error opening file!\n");
		exit(1);
	}
	fprintf(fileptrconfig,  "{\n");

	// Write runtime parameters
	fprintf(fileptrconfig, "\"runtime_param\": {\n");
	fprintf(fileptrconfig, "\"max_iterations\":%d,\n", run_param.max_iterations);
	fprintf(fileptrconfig, "\"convergence_window\":%d,\n", run_param.convergence_window);
	fprintf(fileptrconfig, "\"convergence_threshold\":%f\n", run_param.convergence_threshold);
	fprintf(fileptrconfig, "},\n");

	// Config GA param
	fprintf(fileptrconfig, "\"gene_pool_param\":{\n");
	fprintf(fileptrconfig, "\"genes\":%d,\n", gene_pool.genes);
	fprintf(fileptrconfig, "\"individuals\":%d,\n", gene_pool.individuals);
	fprintf(fileptrconfig, "\"elitism\":%d\n", gene_pool.elitism);
	fprintf(fileptrconfig, "},\n");

	// selection param
	fprintf(fileptrconfig, "\"selection_param\":{\n");
	fprintf(fileptrconfig, "\"selection_method\":%d\n,", config_ga.selection_param.selection_method);
	fprintf(fileptrconfig, "\"selection_div_param\":%f\n,", config_ga.selection_param.selection_div_param);
	fprintf(fileptrconfig, "\"selection_prob_param\":%f\n,", config_ga.selection_param.selection_prob_param);
	fprintf(fileptrconfig, "\"selection_temp_param\":%f\n,", config_ga.selection_param.selection_temp_param);
	fprintf(fileptrconfig, "\"selection_tournament_size\":%d\n", config_ga.selection_param.selection_tournament_size);
	fprintf(fileptrconfig, "},\n");

	// flattem param
	fprintf(fileptrconfig, "\"flatten_param\":{\n");
	fprintf(fileptrconfig, "\"flatten_method\":%d\n,", config_ga.flatten_param.flatten_method);
	fprintf(fileptrconfig, "\"flatten_factor\":%f\n,", config_ga.flatten_param.flatten_factor);
	fprintf(fileptrconfig, "\"flatten_bias\":%f\n,", config_ga.flatten_param.flatten_bias);
	fprintf(fileptrconfig, "\"flatten_optim_mode\":%d\n", config_ga.flatten_param.flatten_optim_mode);
	fprintf(fileptrconfig, "},\n");

	// crossover param
	fprintf(fileptrconfig, "\"crossover_param\":{\n");
	fprintf(fileptrconfig, "\"crossover_method\":%d,\n", config_ga.crossover_param.crossover_method);
	fprintf(fileptrconfig, "\"crossover_prob\":%f\n", config_ga.crossover_param.crossover_prob);
	fprintf(fileptrconfig, "},\n");


	// mutation param
	fprintf(fileptrconfig, "\"mutation_param\":{\n");
	fprintf(fileptrconfig, "\"mutation_method\":%d,\n", config_ga.mutation_param.mutation_method);
	fprintf(fileptrconfig, "\"mutation_prob\":%f,\n", config_ga.mutation_param.mutation_prob);
	fprintf(fileptrconfig, "\"mutation_rate\":%d\n", config_ga.mutation_param.mutation_rate);
	fprintf(fileptrconfig, "},\n");

	// fx param
	fprintf(fileptrconfig, "\"fx_param\":{\n");
	fprintf(fileptrconfig, "\"fx_method\":%d,\n", config_ga.fx_param.fx_method);
	fprintf(fileptrconfig, "\"fx_optim_mode\":%d,\n", config_ga.fx_param.fx_optim_mode);
	fprintf(fileptrconfig, "\"bin2double_factor\":%f,\n", config_ga.fx_param.bin2double_factor);
	fprintf(fileptrconfig, "\"bin2double_bias\":%f\n", config_ga.fx_param.bin2double_bias);
	fprintf(fileptrconfig, "}\n");

	fprintf(fileptrconfig, "}\n");
}

File: src/Logger/logging.py

from dataclasses import dataclass
import json
import struct
from typing import List
from weakref import ProxyType
import numpy as np
 
@dataclass
class runtime_param_s:
    max_iterations: int = 0
    convergence_window: int = 0
    convergence_threshold: float = 0.0
    
@dataclass
class gene_pool_param_s:
    genes: int = 0
    individuals: int = 0
    elitism: int = 0
    
@dataclass
class selection_param_s:
    selection_method: int = 0
    selection_div_param: float = 0.0
    selection_prob_param: float = 0.0
    selection_temp_param: float = 0.0
    selection_tournament_size: int = 0
    
@dataclass
class flatten_param_s:
    flatten_method: int = 0
    flatten_factor: float = 0.0
    flatten_bias: float = 0.0
    flatten_optim_mode: int = 0

@dataclass
class crossover_param_s:
    crossover_method: int = 0
    crossover_prob: float = 0.0

@dataclass
class mutate_param_s:
    mutation_method: int = 0
    mutation_prob: float = 0.0
    mutation_rate: float = 0.0
    
@dataclass
class fx_param_s:
    fx_method: int = 0
    fx_optim_mode: float = 0.0

@dataclass
class population_param_s:
    pop_sampling_type: int = 0
    pop_sigma: int = 0
    pop_lower: List[float] = None
    pop_upper: List[float] = None



class Paramset:
    def __init__(self, fully_qualified_basename: str):
        conifg_dict: dict = json.load(open(fully_qualified_basename, mode="r"))
        self.runtime_param = runtime_param_s(**conifg_dict["runtime_param"])
        self.gene_pool_param = gene_pool_param_s(**conifg_dict["gene_pool_param"])
        self.selection_param = selection_param_s(**conifg_dict["selection_param"])
        self.flatten_param = flatten_param_s(**conifg_dict["flatten_param"])
        self.crossover_param = crossover_param_s(**conifg_dict["crossover_param"])
        self.mutate_param = mutate_param_s(**conifg_dict["mutation_param"])
        self.fx_param = fx_param_s(**conifg_dict["fx_param"])
        self.population_param = population_param_s(**conifg_dict["population_param"])


class Dataset:
    def __init__(self, fully_qualified_basename) -> None:
        self.paramset = Paramset(fully_qualified_basename+".json")
        self.fqb = fully_qualified_basename
        
        self.iterations = np.array([])
        self.resultset = np.array([])
        self.popparamdouble = np.array([])
        
    def read_data(self):
        
        sizeof_iteration = 4
        sizeof_reseultset = 8 
        sizeof_popparamdouble = 8 * self.paramset.gene_pool_param.genes 
        
        file = open(self.fqb + ".bin", mode="rb")
        
        data = file.read()
        
        current_byte = 0
        current_ind = 0
        while current_byte < len(data):
            self.iterations = np.append(self.iterations, struct.unpack("i", data[current_byte:current_byte+sizeof_iteration]))
            current_byte += sizeof_iteration
            
            current_ind = 0
            while current_ind < self.paramset.gene_pool_param.individuals:
                self.resultset = np.append(self.resultset, struct.unpack("d", data[current_byte:current_byte+sizeof_reseultset]))
                current_byte += sizeof_reseultset
            
                self.popparamdouble = np.append(self.popparamdouble, struct.unpack("d"*self.paramset.gene_pool_param.genes, data[current_byte:current_byte+sizeof_popparamdouble]))
                current_byte += sizeof_popparamdouble
                current_ind += 1
            

        self.resultset = self.resultset.reshape((len(self.iterations), self.paramset.gene_pool_param.individuals))
        self.popparamdouble = self.popparamdouble.reshape((len(self.iterations), self.paramset.gene_pool_param.individuals, self.paramset.gene_pool_param.genes))
            



            
if __name__ == "__main__":
    ds = dataset("C:\\temp\\GAThread0")
    ds.read_data()
    print(ds.iterations, len(ds.iterations))
    print(ds.resultset)
    print(ds.popparamdouble.shape)